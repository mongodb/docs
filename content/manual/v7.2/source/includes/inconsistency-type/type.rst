
Type of inconsistency document.  Returns a value of |incon-type| for this   
inconsistency type.


