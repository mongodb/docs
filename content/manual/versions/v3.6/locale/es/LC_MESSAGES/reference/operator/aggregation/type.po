# SOME DESCRIPTIVE TITLE.
# Copyright (C) 2008-2019
# This file is distributed under the same license as the mongodb-manual
# package.
# FIRST AUTHOR <EMAIL@ADDRESS>, 2019.
#
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: mongodb-manual 3.6\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2019-03-19 11:30-0400\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=utf-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Generated-By: Babel 2.6.0\n"

#: ../source/reference/operator/aggregation/type.txt:3
msgid "$type (aggregation)"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:14
msgid "Definition"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:20
msgid ""
"Returns a string that specifies the :doc:`BSON type </reference/bson-"
"types/>` of the argument."
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:23
msgid ""
":expression:`$type` has the following :ref:`operator expression syntax "
"<agg-quick-ref-operator-expressions>`:"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:30
msgid "The argument can be any valid :ref:`expression <aggregation-expressions>`."
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:35
msgid ""
"If you wish to filter documents by their :ref:`BSON type <bson-types>` "
"rather than returning the type of an expression, use the :query:`$type` "
"query operator."
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:40
msgid "Behavior"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:43
msgid "``$type``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:45
msgid ""
"Unlike the :query:`$type` query operator, which matches array elements "
"based on their BSON type, the :expression:`$type` aggregation operator "
"does not examine array elements. Instead, when passed an array as its "
"argument, the ``$type`` aggregation operator returns the type of the "
"argument, i.e. ``\"array\"``."
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:51
#: ../source/reference/operator/aggregation/type.txt:101
msgid ""
"If the argument is a field that is missing in the input document, "
":expression:`$type` returns the string ``\"missing\"``."
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:54
msgid ""
"The following table shows the :expression:`$type` output for several "
"common types of expressions:"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:61
#: ../source/reference/operator/aggregation/type.txt:105
msgid "Example"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:62
msgid "Results"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:64
msgid "``{ $type: \"a\" }``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:65
msgid "``\"string\"``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:67
msgid "``{ $type: /a/ }``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:68
msgid "``\"regex\"``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:70
msgid "``{ $type: 1 }``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:71
msgid "``\"double\"``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:73
msgid "``{ $type: NumberLong(627) }``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:74
msgid "``\"long\"``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:76
msgid "``{ $type: { x: 1 } }``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:77
msgid "``\"object\"``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:79
msgid "``{ $type: [ [ 1, 2, 3 ] ] }``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:80
msgid "``\"array\"``"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:84
msgid ""
"In the case of a literal array such as ``[ 1, 2, 3 ]``, enclose the "
"expression in an outer set of array brackets to prevent MongoDB from "
"parsing ``[ 1, 2, 3 ]`` as an :ref:`argument list <agg-quick-ref-"
"operator-expressions>` with three arguments (``1, 2, 3``). Wrapping the "
"array ``[ 1, 2, 3 ]`` in a :expression:`$literal` expression achieves the"
" same result."
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:93
msgid ""
"See :ref:`operator expression syntax forms <agg-quick-ref-operator-"
"expressions>` for more information."
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:97
msgid "Available Types"
msgstr ""

#: ../source/includes/fact-bson-types.rst:4
msgid "Type"
msgstr ""

#: ../source/includes/fact-bson-types.rst:5
msgid "Number"
msgstr ""

#: ../source/includes/fact-bson-types.rst:6
msgid "Alias"
msgstr ""

#: ../source/includes/fact-bson-types.rst:7
msgid "Notes"
msgstr ""

#: ../source/includes/fact-bson-types.rst:9
msgid "Double"
msgstr ""

#: ../source/includes/fact-bson-types.rst:10
msgid "1"
msgstr ""

#: ../source/includes/fact-bson-types.rst:11
msgid "\"double\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:14
msgid "String"
msgstr ""

#: ../source/includes/fact-bson-types.rst:15
msgid "2"
msgstr ""

#: ../source/includes/fact-bson-types.rst:16
msgid "\"string\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:19
msgid "Object"
msgstr ""

#: ../source/includes/fact-bson-types.rst:20
msgid "3"
msgstr ""

#: ../source/includes/fact-bson-types.rst:21
msgid "\"object\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:24
msgid "Array"
msgstr ""

#: ../source/includes/fact-bson-types.rst:25
msgid "4"
msgstr ""

#: ../source/includes/fact-bson-types.rst:26
msgid "\"array\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:29
msgid "Binary data"
msgstr ""

#: ../source/includes/fact-bson-types.rst:30
msgid "5"
msgstr ""

#: ../source/includes/fact-bson-types.rst:31
msgid "\"binData\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:34
msgid "Undefined"
msgstr ""

#: ../source/includes/fact-bson-types.rst:35
msgid "6"
msgstr ""

#: ../source/includes/fact-bson-types.rst:36
msgid "\"undefined\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:37
#: ../source/includes/fact-bson-types.rst:67
#: ../source/includes/fact-bson-types.rst:77
msgid "Deprecated."
msgstr ""

#: ../source/includes/fact-bson-types.rst:39
msgid "ObjectId"
msgstr ""

#: ../source/includes/fact-bson-types.rst:40
msgid "7"
msgstr ""

#: ../source/includes/fact-bson-types.rst:41
msgid "\"objectId\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:44
msgid "Boolean"
msgstr ""

#: ../source/includes/fact-bson-types.rst:45
msgid "8"
msgstr ""

#: ../source/includes/fact-bson-types.rst:46
msgid "\"bool\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:49
msgid "Date"
msgstr ""

#: ../source/includes/fact-bson-types.rst:50
msgid "9"
msgstr ""

#: ../source/includes/fact-bson-types.rst:51
msgid "\"date\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:54
msgid "Null"
msgstr ""

#: ../source/includes/fact-bson-types.rst:55
msgid "10"
msgstr ""

#: ../source/includes/fact-bson-types.rst:56
msgid "\"null\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:59
msgid "Regular Expression"
msgstr ""

#: ../source/includes/fact-bson-types.rst:60
msgid "11"
msgstr ""

#: ../source/includes/fact-bson-types.rst:61
msgid "\"regex\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:64
msgid "DBPointer"
msgstr ""

#: ../source/includes/fact-bson-types.rst:65
msgid "12"
msgstr ""

#: ../source/includes/fact-bson-types.rst:66
msgid "\"dbPointer\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:69
msgid "JavaScript"
msgstr ""

#: ../source/includes/fact-bson-types.rst:70
msgid "13"
msgstr ""

#: ../source/includes/fact-bson-types.rst:71
msgid "\"javascript\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:74
msgid "Symbol"
msgstr ""

#: ../source/includes/fact-bson-types.rst:75
msgid "14"
msgstr ""

#: ../source/includes/fact-bson-types.rst:76
msgid "\"symbol\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:79
msgid "JavaScript (with scope)"
msgstr ""

#: ../source/includes/fact-bson-types.rst:80
msgid "15"
msgstr ""

#: ../source/includes/fact-bson-types.rst:81
msgid "\"javascriptWithScope\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:84
msgid "32-bit integer"
msgstr ""

#: ../source/includes/fact-bson-types.rst:85
msgid "16"
msgstr ""

#: ../source/includes/fact-bson-types.rst:86
msgid "\"int\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:89
msgid "Timestamp"
msgstr ""

#: ../source/includes/fact-bson-types.rst:90
msgid "17"
msgstr ""

#: ../source/includes/fact-bson-types.rst:91
msgid "\"timestamp\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:94
msgid "64-bit integer"
msgstr ""

#: ../source/includes/fact-bson-types.rst:95
msgid "18"
msgstr ""

#: ../source/includes/fact-bson-types.rst:96
msgid "\"long\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:99
msgid "Decimal128"
msgstr ""

#: ../source/includes/fact-bson-types.rst:100
msgid "19"
msgstr ""

#: ../source/includes/fact-bson-types.rst:101
msgid "\"decimal\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:102
msgid "New in version 3.4."
msgstr ""

#: ../source/includes/fact-bson-types.rst:104
msgid "Min key"
msgstr ""

#: ../source/includes/fact-bson-types.rst:105
msgid "-1"
msgstr ""

#: ../source/includes/fact-bson-types.rst:106
msgid "\"minKey\""
msgstr ""

#: ../source/includes/fact-bson-types.rst:109
msgid "Max key"
msgstr ""

#: ../source/includes/fact-bson-types.rst:110
msgid "127"
msgstr ""

#: ../source/includes/fact-bson-types.rst:111
msgid "\"maxKey\""
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:107
msgid ""
"This example uses a :term:`collection <collection>` named ``coll`` with "
"the following :term:`documents <document>`:"
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:119
msgid ""
"The following aggregation operation uses the ``$type`` operator to "
"display the type of field ``a`` for all documents as part of the "
":pipeline:`$project` stage."
msgstr ""

#: ../source/reference/operator/aggregation/type.txt:131
msgid "The operation returns the following:"
msgstr ""

