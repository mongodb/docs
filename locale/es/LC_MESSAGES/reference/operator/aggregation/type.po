# SOME DESCRIPTIVE TITLE.
# Copyright (C) 2008-2016
# This file is distributed under the same license as the mongodb-manual
# package.
# FIRST AUTHOR <EMAIL@ADDRESS>, 2019.
#
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: mongodb-manual 3.4\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2016-12-08 12:02-0500\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=utf-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Generated-By: Babel 2.6.0\n"

# d4bbf988f7974f2483b5728ffd72c3bf
#: ../source/reference/operator/aggregation/type.txt:3
msgid "$type (aggregation)"
msgstr ""

# b64bd8c1692c4eea99f174e663e70172
#: ../source/reference/operator/aggregation/type.txt
msgid "On this page"
msgstr ""

# e192009c6d8d4f9eb3dd791386a8525e
#: ../source/reference/operator/aggregation/type.txt:14
msgid "Definition"
msgstr ""

# 7d402f99fec8447ba373533442270be5
#: ../source/reference/operator/aggregation/type.txt:20
msgid ""
"Returns a string that specifies the :doc:`BSON type </reference/bson-"
"types/>` of the argument."
msgstr ""

# f945fd8c35f04e5aae6685fce1744a62
#: ../source/reference/operator/aggregation/type.txt:23
msgid ""
":expression:`$type` has the following :ref:`operator expression syntax "
"<agg-quick-ref-operator-expressions>`:"
msgstr ""

# 588f2206f51b43a5836b5161aa92cc0e
#: ../source/reference/operator/aggregation/type.txt:30
msgid "The argument can be any valid :ref:`expression <aggregation-expressions>`."
msgstr ""

# cc6ad411c27d468d97006b1a03f6cd20
#: ../source/reference/operator/aggregation/type.txt:34
msgid "Behavior"
msgstr ""

# b0cfa8014aac45ba812f016c4f0ae09c
#: ../source/reference/operator/aggregation/type.txt:36
msgid ""
"Unlike the :query:`$type` query operator, which matches array elements "
"based on their BSON type, the :expression:`$type` aggregation operator "
"does not examine array elements. Instead, when passed an array as its "
"argument, the ``$type`` aggregation operator returns the type of the "
"argument, i.e. ``\"array\"``."
msgstr ""

# 22f64d3123a3458a9400f3d2226e30b2
#: ../source/reference/operator/aggregation/type.txt:42
msgid ""
"If the argument is a field that is missing in the input document, "
":expression:`$type` returns the string ``\"missing\"``."
msgstr ""

# 3e4e567ed44a4eb99009492a9464a867
# bd517e8c1ac048f68984aa7b059b2a5e
#: ../source/reference/operator/aggregation/type.txt:53
#: ../source/reference/operator/aggregation/type.txt:89
msgid "Example"
msgstr ""

# 6bc0fb9c1be7482dbb2742b74c550a4d
#: ../source/reference/operator/aggregation/type.txt:54
msgid "Results"
msgstr ""

# ce53f805dce046ba9fadfbc807a17762
#: ../source/reference/operator/aggregation/type.txt:56
msgid "``{ $type: \"a\" }``"
msgstr ""

# eae4710260ac4f8e9ff0f2229752766c
#: ../source/reference/operator/aggregation/type.txt:57
msgid "``\"string\"``"
msgstr ""

# 66f62df1c9a14c01848db22f2c103136
#: ../source/reference/operator/aggregation/type.txt:59
msgid "``{ $type: /a/ }``"
msgstr ""

# 780556e0e1554dedbe1327da6ba07595
#: ../source/reference/operator/aggregation/type.txt:60
msgid "``\"regex\"``"
msgstr ""

# 4c60f0da177842f18367b5c7559507a6
#: ../source/reference/operator/aggregation/type.txt:62
msgid "``{ $type: 1 }``"
msgstr ""

# 1d3a927d71a94a06b4915840248d7a47
#: ../source/reference/operator/aggregation/type.txt:63
msgid "``\"double\"``"
msgstr ""

# fbd0acd7b10d45f4bcadf4b541747e2a
#: ../source/reference/operator/aggregation/type.txt:65
msgid "``{ $type: NumberLong(627) }``"
msgstr ""

# 0a5784fcb9e84d30b51cccab482a9a56
#: ../source/reference/operator/aggregation/type.txt:66
msgid "``\"long\"``"
msgstr ""

# 1f9af95d830c46c98a56437aafde2fe4
#: ../source/reference/operator/aggregation/type.txt:68
msgid "``{ $type: { x: 1 } }``"
msgstr ""

# d31c5584126f4fca8051223ac97ba04a
#: ../source/reference/operator/aggregation/type.txt:69
msgid "``\"object\"``"
msgstr ""

# 7246de5f63514059b431ce73400c951c
#: ../source/reference/operator/aggregation/type.txt:71
msgid "``{ $type: [ [ 1, 2, 3 ] ] }``"
msgstr ""

# b10cb37f321a45aca88ef60ed5c296ef
#: ../source/reference/operator/aggregation/type.txt:72
msgid "``\"array\"``"
msgstr ""

# eb2195ed901d4c3897c761dbca618177
#: ../source/reference/operator/aggregation/type.txt:76
msgid ""
"In the case of a literal array such as ``[ 1, 2, 3 ]``, enclose the "
"expression in an outer set of array brackets to prevent MongoDB from "
"parsing ``[ 1, 2, 3 ]`` as an :ref:`argument list <agg-quick-ref-"
"operator-expressions>` with three arguments (``1, 2, 3``). Wrapping the "
"array ``[ 1, 2, 3 ]`` in a :expression:`$literal` expression achieves the"
" same result."
msgstr ""

# 2eb5e505de2240d388cc7fdb1ac7adea
#: ../source/reference/operator/aggregation/type.txt:85
msgid ""
"See :ref:`operator expression syntax forms <agg-quick-ref-operator-"
"expressions>` for more information."
msgstr ""

# 4234818fa20a46328f7e67ebc05f2177
#: ../source/reference/operator/aggregation/type.txt:91
msgid ""
"This example uses a :term:`collection <collection>` named ``coll`` with "
"the following :term:`documents <document>`:"
msgstr ""

# c0eeac8b6795478988661a9c885d1db7
#: ../source/reference/operator/aggregation/type.txt:103
msgid ""
"The following aggregation operation uses the ``$type`` operator to "
"display the type of field ``a`` for all documents as part of the "
":pipeline:`$project` stage."
msgstr ""

# e5ac12ef5757452eb4e348cde61611df
#: ../source/reference/operator/aggregation/type.txt:115
msgid "The operation returns the following:"
msgstr ""

