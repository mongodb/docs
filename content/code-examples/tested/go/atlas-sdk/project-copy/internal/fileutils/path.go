package fileutils

import (
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// GenerateOutputPath constructs a valid file path based on the given directory, prefix, and optional extension.
// It returns the full path to the generated file with formatted filename or an error if any operation fails.
//
// NOTE: You can define a default global directory for all generated files by setting the ATLAS_DOWNLOADS_DIR environment variable.
func GenerateOutputPath(dir, prefix, extension string) (string, error) {
	// If default download directory is set in .env, prepend it to the provided dir
	dir = ResolveWithDownloadsBase(dir)

	// Create directory if it doesn't exist
	if err := os.MkdirAll(dir, 0755); err != nil {
		return "", fmt.Errorf("failed to create output directory: %w", err)
	}

	timestamp := time.Now().Format("20060102")
	var filename string
	if extension == "" {
		filename = fmt.Sprintf("%s_%s", prefix, timestamp)
	} else {
		filename = fmt.Sprintf("%s_%s.%s", prefix, timestamp, extension)
	}

	filename = filepath.Clean(filename)
	if len(filename) > 255 {
		return "", fmt.Errorf("filename exceeds maximum length of 255 characters: %s", filename)
	}

	return filepath.Join(dir, filename), nil
}
