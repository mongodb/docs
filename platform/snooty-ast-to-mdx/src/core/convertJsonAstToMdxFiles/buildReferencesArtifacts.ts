import fs from 'node:fs/promises';

type Substitutions = Record<string, string>;
type Refs = Record<string, { title: string; url: string }>;

export interface ReferencesArtifact {
  substitutions: Substitutions;
  refs: Refs;
}

/** Build references artifacts (TS and JSON) for a given set of substitutions and refs */
export const buildReferencesArtifacts = ({
  substitutions = {},
  refs = {},
}: ReferencesArtifact): { ts: string; json: string } => {
  const subsLines = Object.entries(substitutions)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([k, v]) => `    ${JSON.stringify(k)}: ${esc(v)},`)
    .join('\n');

  const refsLines = Object.entries(refs)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([url, { title }]) => `    ${esc(url)}: { title: ${esc(title)}, url: ${esc(url)} },`)
    .join('\n');

  const ts =
    `export const substitutions = {\n${subsLines}\n} as const;\n\n` +
    `export const refs = {\n${refsLines}\n} as const;\n\n` +
    `const references = { substitutions, refs } as const;\n\n` +
    `export default references;\n`;

  const json = JSON.stringify({ substitutions, refs }, null, 2) + '\n';

  return { ts, json };
};

export const readExistingReferences = async (tsFilePath: string): Promise<ReferencesArtifact> => {
  const jsonPath = tsFilePath.replace(/\.ts$/, '.json');
  try {
    const buffer = await fs.readFile(jsonPath, 'utf8');
    const existing = JSON.parse(buffer);
    const { substitutions, refs } = existing ?? {};
    return { substitutions, refs };
  } catch {
    return { substitutions: {}, refs: {} };
  }
};

interface MergeReferencesArgs {
  base: ReferencesArtifact;
  add: ReferencesArtifact;
}

export const mergeReferences = ({ base, add }: MergeReferencesArgs): ReferencesArtifact => ({
  substitutions: { ...base.substitutions, ...(add.substitutions ?? {}) },
  refs: { ...base.refs, ...(add.refs ?? {}) },
});

const esc = (value: string) => JSON.stringify(String(value).slice(0, 1_000));
