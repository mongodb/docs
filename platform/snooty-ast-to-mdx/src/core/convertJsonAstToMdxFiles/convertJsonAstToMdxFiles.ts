import fs from 'node:fs/promises';
import { posix as path } from 'node:path';
import { convertMdastToMdx } from '../convertMdastToMdx';
import { convertSnootyAstToMdast } from '../convertSnootyAstToMdast/convertSnootyAstToMdast';
import { buildReferencesArtifacts, readExistingReferences, mergeReferences } from './buildReferencesArtifacts';
import type { ReferencesArtifact } from './buildReferencesArtifacts';
import type { SnootyNode } from '../convertSnootyAstToMdast/types';

interface ConvertJsonAstToMdxFilesOptions {
  ast: SnootyNode;
  outputPath: string;
  outputRootDir?: string;
}

type ConvertJsonAstToMdxFiles = (args: ConvertJsonAstToMdxFilesOptions) => Promise<{
  fileCount: number;
  emittedReferencesFile?: string;
}>;

/** Convert a JSON AST to an MDX file and create additional files (refs, includes, etc.) */
export const convertJsonAstToMdxFiles: ConvertJsonAstToMdxFiles = async ({ ast, outputPath, outputRootDir }) => {
  // Track unique output file paths we've written during this process to avoid
  // duplicate writes and over-counting when includes repeat across pages.
  const emittedFilePaths = new Set<string>();
  const additionalRefs: ReferencesArtifact = {
    substitutions: {},
    refs: {},
  };

  const rootDir = outputRootDir ?? path.dirname(outputPath);
  const mdast = convertSnootyAstToMdast(ast, {
    // Make the current output file path relative to the provided output root directory
    currentOutfilePath: path.relative(rootDir, outputPath),
    onEmitMdxFile: async ({ outfilePath, mdastRoot }) => {
      try {
        const outPath = path.join(rootDir, outfilePath);
        const resolvedOutPath = path.resolve(outPath);

        if (!emittedFilePaths.has(resolvedOutPath)) {
          const mdxContent = convertMdastToMdx(mdastRoot);

          await fs.mkdir(path.dirname(outPath), { recursive: true });
          await fs.writeFile(outPath, mdxContent);
          emittedFilePaths.add(resolvedOutPath);
        }

        const references = mdastRoot.__references as ReferencesArtifact;
        if (references) {
          Object.assign(additionalRefs.substitutions, references.substitutions || {});
          Object.assign(additionalRefs.refs, references.refs || {});
        }
      } catch (error) {
        console.error('Failed to emit include file:', outfilePath, error);
      }
    },
  });

  const resolvedMainOutPath = path.resolve(outputPath);

  if (!emittedFilePaths.has(resolvedMainOutPath)) {
    const mdx = convertMdastToMdx(mdast);

    await fs.mkdir(path.dirname(outputPath), { recursive: true });
    await fs.writeFile(outputPath, mdx);
    emittedFilePaths.add(resolvedMainOutPath);
  }

  const refsArtifact = mdast.__references as ReferencesArtifact;
  const emittedReferencesFile = await createReferencesFiles({ rootDir, refsArtifact, additionalRefs });

  return {
    fileCount: emittedFilePaths.size,
    emittedReferencesFile,
  };
};

interface CreateReferencesFilesArgs {
  rootDir: string;
  refsArtifact: ReferencesArtifact;
  additionalRefs: ReferencesArtifact;
}

/** If references were collected, emit or update a references.ts file at the output root directory */
const createReferencesFiles = async ({ rootDir, refsArtifact, additionalRefs }: CreateReferencesFilesArgs) => {
  if (refsArtifact || Object.keys(additionalRefs.substitutions).length || Object.keys(additionalRefs.refs).length) {
    if (refsArtifact) {
      Object.assign(additionalRefs.substitutions, refsArtifact.substitutions || {});
      Object.assign(additionalRefs.refs, refsArtifact.refs || {});
    }

    const refsPath = path.join(rootDir, 'references.ts');
    await fs.mkdir(path.dirname(refsPath), { recursive: true });

    const existing = await readExistingReferences(refsPath);
    const merged = mergeReferences({ base: existing, add: additionalRefs });
    const { ts: tsOut, json: jsonOut } = buildReferencesArtifacts(merged);

    await fs.writeFile(refsPath, tsOut);
    await fs.writeFile(refsPath.replace(/\.ts$/, '.json'), jsonOut);
    return refsPath;
  }
};
