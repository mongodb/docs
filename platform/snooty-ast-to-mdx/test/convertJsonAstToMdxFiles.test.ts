import fs from 'node:fs/promises';
import path from 'node:path';
import type { SnootyNode } from '../src/core/convertSnootyAstToMdast/types';
import { convertJsonAstToMdxFiles } from '../src/core/convertJsonAstToMdxFiles/convertJsonAstToMdxFiles';

// no files will get output since we mock the fs calls below
const OUTPUT_PATH = './test/__snapshots__/out.mdx';
// however we can get the resulting MDX strings from the mock calls (for snapshots, etc.)
const fsWriteFileSpy = jest.spyOn(fs, 'writeFile').mockImplementation(() => Promise.resolve());
const fsMkdirSpy = jest.spyOn(fs, 'mkdir').mockImplementation(() => Promise.resolve(undefined));
const fsReadFileSpy = jest.spyOn(fs, 'readFile').mockImplementation(() => Promise.resolve(''));

const simpleAst: SnootyNode = {
  type: 'root',
  children: [
    {
      type: 'paragraph',
      children: [{ type: 'text', value: 'Body' }],
    },
  ],
};

describe('convertJsonAstToMdx', () => {
  it('writes only the main file when there are no includes or references', async () => {
    const { fileCount, emittedReferencesFile } = await convertJsonAstToMdxFiles({
      ast: simpleAst,
      outputPath: OUTPUT_PATH,
    });

    expect(fileCount).toBe(1);
    expect(emittedReferencesFile).toBeUndefined();

    expect(fsWriteFileSpy).toHaveBeenCalledWith(OUTPUT_PATH, expect.any(String));
    expect(fsMkdirSpy).toHaveBeenCalledWith(path.dirname(OUTPUT_PATH), { recursive: true });
    expect(fsReadFileSpy).not.toHaveBeenCalled();

    const writtenMdx = (fsWriteFileSpy.mock.calls[0][1] as string).trim();
    expect(writtenMdx).toMatchSnapshot('simple-no-refs');
  });

  it('emits include file only once even if referenced twice', async () => {
    const includedRstPath = 'included-file.rst';
    const ast: SnootyNode = {
      type: 'root',
      children: [
        {
          type: 'directive',
          name: 'include',
          argument: includedRstPath,
          children: [
            {
              type: 'paragraph',
              children: [{ type: 'text', value: 'Hi from include' }],
            },
          ],
        },
        // duplicate include
        {
          type: 'directive',
          name: 'include',
          argument: includedRstPath,
          children: [
            {
              type: 'paragraph',
              children: [{ type: 'text', value: 'Hi from include' }],
            },
          ],
        },
      ],
    };

    const { fileCount } = await convertJsonAstToMdxFiles({ ast, outputPath: OUTPUT_PATH });

    expect(fileCount).toBe(2);

    const includePath = path.join(path.dirname(OUTPUT_PATH), includedRstPath.replace('.rst', '.mdx'));
    expect(fsWriteFileSpy).toHaveBeenCalledWith(includePath, expect.any(String));
    expect(fsMkdirSpy).toHaveBeenCalledWith(path.dirname(OUTPUT_PATH), { recursive: true });
    expect(fsReadFileSpy).not.toHaveBeenCalled();

    const writes = Object.fromEntries(
      fsWriteFileSpy.mock.calls.map(([path, content]) => [path, String(content).trim()]),
    );
    expect(writes[OUTPUT_PATH]).toMatchSnapshot('main-mdx-with-include');
    expect(writes[includePath]).toMatchSnapshot('include-mdx');
  });

  it('collects substitutions/refs and emits references artifacts', async () => {
    const ast: SnootyNode = {
      type: 'root',
      children: [
        {
          type: 'paragraph',
          children: [
            {
              type: 'substitution_reference',
              refname: 'prod',
              children: [{ type: 'text', value: 'MongoDB' }],
            },
            {
              type: 'doc',
              url: '/docs/page',
              children: [{ type: 'text', value: 'Doc Page' }],
            },
          ],
        },
      ],
    };

    const { fileCount, emittedReferencesFile } = await convertJsonAstToMdxFiles({ ast, outputPath: OUTPUT_PATH });

    expect(fileCount).toBe(1);
    expect(emittedReferencesFile).toBeDefined();

    const refsJson = emittedReferencesFile!.replace('.ts', '.json');
    expect(fsWriteFileSpy).toHaveBeenCalledWith(emittedReferencesFile, expect.any(String));
    expect(fsWriteFileSpy).toHaveBeenCalledWith(refsJson, expect.any(String));
    expect(fsMkdirSpy).toHaveBeenCalledWith(path.dirname(OUTPUT_PATH), { recursive: true });
    expect(fsReadFileSpy).toHaveBeenCalled();

    const jsonEntry = fsWriteFileSpy.mock.calls.find(([path]) => path === refsJson) as [string, string];
    expect(jsonEntry).toBeDefined();
    expect(JSON.parse(jsonEntry[1])).toMatchSnapshot('references-json');
  });

  it('respects outputRootDir when emitting includes', async () => {
    const rootDir = './tmp/out';
    const mainOut = path.join(rootDir, 'guides/page.mdx');

    const sharedRstPath = 'shared/included.rst';
    const ast: SnootyNode = {
      type: 'root',
      children: [
        {
          type: 'directive',
          name: 'include',
          argument: sharedRstPath,
          children: [
            {
              type: 'paragraph',
              children: [{ type: 'text', value: 'some included content' }],
            },
          ],
        },
      ],
    };

    await convertJsonAstToMdxFiles({ ast, outputPath: mainOut, outputRootDir: rootDir });

    const mdxIncludedPath = path.join(rootDir, sharedRstPath.replace('.rst', '.mdx'));
    expect(fsWriteFileSpy).toHaveBeenCalledWith(mdxIncludedPath, expect.any(String));
    expect(fsMkdirSpy).toHaveBeenCalledWith(path.dirname(OUTPUT_PATH), { recursive: true });
    expect(fsReadFileSpy).toHaveBeenCalled();
  });

  it('will throw an error if the writeFile fails', async () => {
    const errorMessage = 'write error';
    fsWriteFileSpy.mockImplementationOnce(() => Promise.reject(errorMessage));

    try {
      await convertJsonAstToMdxFiles({ ast: simpleAst, outputPath: OUTPUT_PATH });

      fail('Expected an error to be thrown');
    } catch (error) {
      expect(error).toBe(errorMessage);
    }
  });
});
