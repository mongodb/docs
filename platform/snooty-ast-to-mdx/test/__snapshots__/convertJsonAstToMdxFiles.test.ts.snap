// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`convertJsonAstToMdx collects substitutions/refs and emits references artifacts: references-json 1`] = `
{
  "refs": {
    "/docs/page": {
      "title": "Doc Page",
      "url": "/docs/page",
    },
  },
  "substitutions": {
    "prod": "MongoDB",
  },
}
`;

exports[`convertJsonAstToMdx emits include file only once even if referenced twice: include-mdx 1`] = `"Hi from include"`;

exports[`convertJsonAstToMdx emits include file only once even if referenced twice: main-mdx-with-include 1`] = `
"import IncludedFile from './included-file.mdx';

<IncludedFile />

<IncludedFile />"
`;

exports[`convertJsonAstToMdx writes only the main file when there are no includes or references: simple-no-refs 1`] = `"Body"`;
